### ISSUE TITLE
[Nullness checker] Type not refined if null check is in and expression

### ISSUE THREAD

**Issue by Ao-senXiong** [eisop/checker-framework#1029]
The following code is a false positive case for nullness checker.

command: ./checker/bin/javac -processor nullness Test.java

source code:
```
import org.checkerframework.checker.nullness.qual.Nullable;
import java.util.List;

public class Test {
    @Nullable List test(){return null;}

    boolean foo() {
        return this.test() != null && this.test().isEmpty();
    }
}
```

```
Test.java:8: error: [dereference.of.nullable] dereference of possibly-null reference this.test()
        return this.test() != null && this.test().isEmpty();
                                               ^
1 error
```

**Comment by Ao-senXiong** [comment]
@wmdietl Should the return type of `test()` be refined to Nonnull for above case? I am wondering do we need to other constraint like `@deterministic` to do the refinement.

**Comment by wmdietl** [comment]
This is not a false positive, this is intended.
The method is not pure, so it could have different behavior for the two invocations.
See the manual on [side effects](https://eisop.github.io/cf/manual/manual.html#type-refinement-purity) and [how to suppress these warnings](https://eisop.github.io/cf/manual/manual.html#purity-suppress-warnings).
